### ISSUE TITLE
xdeps unregister misses targets?

### ISSUE THREAD

**Issue by giadarol** [xsuite/xsuite#368]
I remove all expressions for LHC IR dipole correctors by executing the following

```python 
import xtrack as xt
import lhc_match as lm

default_tol = {None: 1e-8, 'betx': 1e-6, 'bety': 1e-6} # to have no rematching w.r.t. madx

collider = xt.Multiline.from_json('hllhc.json')
collider.build_trackers()
collider.vars.load_madx_optics_file(
    "../../test_data/hllhc15_thick/opt_round_150_1500.madx")

tw0 = collider.twiss()

# Inspect knob expressions
all_knobs_ip2ip8 = ['acbxh3.r2', 'acbchs5.r2b1', 'pxip2b1', 'acbxh2.l8',
    'acbyhs4.r8b2', 'pyip2b1', 'acbxv1.l8', 'acbyvs4.l2b1', 'acbxh1.l8',
    'acbxv2.r8', 'pxip8b2', 'yip8b1', 'pxip2b2', 'acbcvs5.r2b1', 'acbyhs4.l8b1',
    'acbyvs4.l8b1', 'acbxh2.l2', 'acbxh3.l2', 'acbxv1.r8', 'acbxv1.r2',
    'acbyvs4.r2b2', 'acbyvs4.l2b2', 'yip8b2', 'xip2b2', 'acbxh2.r2',
    'acbyhs4.l2b2', 'acbxv2.r2', 'acbyhs5.r8b1', 'acbxh2.r8', 'acbxv3.r8',
    'acbyvs5.r8b2', 'acbyvs5.l2b2', 'yip2b1', 'acbxv2.l2', 'acbyhs4.r2b2',
    'acbyhs4.r2b1', 'xip8b2', 'acbyvs5.l2b1', 'acbyvs4.r8b1', 'acbyvs4.r8b2',
    'acbyvs5.r8b1', 'acbxh1.r8', 'acbyvs4.l8b2', 'acbyhs5.l2b1', 'acbyvs4.r2b1',
    'acbcvs5.r2b2', 'acbcvs5.l8b2', 'acbyhs4.r8b1', 'pxip8b1', 'acbxv1.l2',
    'yip2b2', 'acbyhs4.l8b2', 'acbxv3.r2', 'xip8b1', 'acbchs5.r2b2', 'acbxh3.l8',
    'acbxh3.r8', 'acbyhs5.r8b2', 'acbxv2.l8', 'acbxh1.l2', 'pyip8b1', 'pyip8b2',
    'acbxv3.l8', 'xip2b1', 'acbyhs5.l2b2', 'acbchs5.l8b2', 'acbcvs5.l8b1',
    'pyip2b2', 'acbxv3.l2', 'acbchs5.l8b1', 'acbyhs4.l2b1', 'acbxh1.r2']

# kill all existing knobs
for kk in all_knobs_ip2ip8:
    collider.vars[kk] = 0
```

But I still find some targets attached to crossing angle knobs:
```
In [26]: collider.vars['on_x8h']._info(None)
#  vars['on_x8h']._get_value()
   vars['on_x8h'] = 0.0

#  vars['on_x8h']._expr
   vars['on_x8h'] = (vars['on_x8']*vars['cphi_ir8'])

#  vars['on_x8h']._expr._get_dependencies()
   vars['cphi_ir8'] = 1.0
   vars['on_x8'] = 0.0

#  vars['on_x8h']._find_dependant_targets()
   eref['lhcb1']['mcbyh.5r8.b1'].knl[0]
   eref['lhcb1']['mcbch.b5l8.b1'].knl[0]
   eref['lhcb2']['mcbxh.1r8'].knl[0]
   eref['lhcb1']['mcbxh.1r8'].knl[0]
   eref['lhcb2']['mcbxh.1l8'].knl[0]
   eref['lhcb1']['mcbxh.1l8'].knl[0]
   eref['lhcb2']['mcbxh.2l8'].knl[0]
   eref['lhcb1']['mcbxh.2l8'].knl[0]
   eref['lhcb2']['mcbxh.2r8'].knl[0]
   eref['lhcb1']['mcbxh.2r8'].knl[0]
   eref['lhcb2']['mcbxh.3l8'].knl[0]
   eref['lhcb1']['mcbxh.3l8'].knl[0]
   eref['lhcb2']['mcbxh.3r8'].knl[0]
   eref['lhcb1']['mcbxh.3r8'].knl[0]
   eref['lhcb1']['mcbyh.4l8.b1'].knl[0]
   eref['lhcb1']['mcbyh.a4r8.b1'].knl[0]
   eref['lhcb2']['mcbyh.a4l8.b2'].knl[0]
   eref['lhcb2']['mcbyh.4r8.b2'].knl[0]
   eref['lhcb2']['mcbch.5l8.b2'].knl[0]
   eref['lhcb2']['mcbyh.b5r8.b2'].knl[0]
```

The model consistency is preserved, but some unnecessary updates are triggered.

**Comment by giadarol** [comment]
Also the `rebuild` and the `verify` methods don't seem to work both with and without redefining the knobs.

@szymonlopaciuk Maybe this is related to the recente changes in the `unregister`?

```
In [32]: collider._xdeps_manager.rebuild()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[32], line 1
----> 1 collider._xdeps_manager.rebuild()

File ~/xsuite_packages/xdeps/xdeps/tasks.py:430, in Manager.rebuild(self)
    429 def rebuild(self):
--> 430     self.cleanup()
    431     other = Manager()
    432     other.containers.update(self.containers)

File ~/xsuite_packages/xdeps/xdeps/tasks.py:413, in Manager.cleanup(self)
    411 for dct in self.rdeps, self.rtasks, self.deptasks, self.tartasks:
    412     for kk, ss in list(dct.items()):
--> 413         if len(ss) == 0:
    414             del dct[kk]

TypeError: object of type 'RefList' has no len()

In [33]: collider._xdeps_manager.verify()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[33], line 1
----> 1 collider._xdeps_manager.verify()

File ~/xsuite_packages/xdeps/xdeps/tasks.py:439, in Manager.verify(self, dcts)
    438 def verify(self, dcts=("rdeps", "rtasks", "deptasks", "tartasks")):
--> 439     other = self.rebuild()
    440     for dct in dcts:
    441         odct = getattr(other, dct)

File ~/xsuite_packages/xdeps/xdeps/tasks.py:430, in Manager.rebuild(self)
    429 def rebuild(self):
--> 430     self.cleanup()
    431     other = Manager()
    432     other.containers.update(self.containers)

File ~/xsuite_packages/xdeps/xdeps/tasks.py:413, in Manager.cleanup(self)
    411 for dct in self.rdeps, self.rtasks, self.deptasks, self.tartasks:
    412     for kk, ss in list(dct.items()):
--> 413         if len(ss) == 0:
    414             del dct[kk]

TypeError: object of type 'RefList' has no len()
```

**Comment by szymonlopaciuk** [comment]
It's possible, I'll try to look into it.

**Comment by szymonlopaciuk** [comment]
I thought I found where the problem was earlier, but it seems I confused some knobs... As soon as I started to write a minimal example, I realised I cannot reproduce this. Perhaps I am using a different model, I took mine from `examples/knobs/001_lhc.py`, and I con execute your code with `line` instead of `collider` seemingly without problems.

Perhaps the issue is somehow dependent on your specific example? It seems also that our versions of Xdeps are different, since for me `._info(None)` throws an exception, and I am on the latest main.
